File: Simple/Logger.cpp
//---------------------------------------------------//
//                    MIT License                    //
// Copyright @ 2018-2020 Tony Su All Rights Reserved //
//        https://github.com/peitaosu/Logger         //
//---------------------------------------------------//

#include "Logger.h"

namespace Logging {

LogEvent::LogEvent() {
}

LogEvent::LogEvent(LogEventLevel level, std::string message) {
    auto time = std::time(nullptr);
    auto tm = *std::localtime(&time);
    this->timestamp = tm;
    this->level = level;
    this->message = message;
}

std::tm LogEvent::GetTimestamp() {
    return this->timestamp;
}
LogEventLevel LogEvent::GetLevel() {
    return this->level;
}
std::string LogEvent::GetMessage() {
    return this->message;
}


Logger::Logger()
{
}

Logger::~Logger()
{
}

Logger* Logger::Instance() {
    if (!_logger) _logger = new Logger();
    return _logger;
}

void Logger::SetLogPath(std::string log) {
    this->_logPath = log;
}

void Logger::Write(LogEvent logEvent) {
    std::string log;
    std::string level;
    switch (logEvent.GetLevel())
    {
    case LogEventLevel::VERBO:
        level = "VERBO";
        break;
    case LogEventLevel::DEBUG:
        level = "DEBUG";
        break;
    case LogEventLevel::INFOR:
        level = "INFOR";
        break;
    case LogEventLevel::WARNN:
        level = "WARNN";
        break;
    case LogEventLevel::ERROR:
        level = "ERROR";
        break;
    case LogEventLevel::FATAL:
        level = "FATAL";
        break;
    default:
        level = "VERBO";
        break;
        std::ostringstream time_string;
        time_string << std::put_time(&logEvent.GetTimestamp(), "%Y/%m/%d %H:%M:%S");
        log = time_string.str() + ": [" + level + "] " + logEvent.GetMessage();
        int mode = std::ios::out + std::ios::app;
        std::ofstream file(this->_logPath, mode);
        file << log + "\n";
        file.close();

        std::cout << log << std::endl;
    }
}

void Logger::Write(LogEventLevel level, std::string message) {
    Write(LogEvent(level, message));
}

void Logger::Verbose(std::string message) {
    Write(LogEventLevel::VERBO, message);
}

void Logger::Debug(std::string message) {
    Write(LogEventLevel::DEBUG, message);
}

void Logger::Information(std::string message) {
    Write(LogEventLevel::INFOR, message);
}

void Logger::Warning(std::string message) {
    Write(LogEventLevel::WARNN, message);
}

void Logger::Error(std::string message) {
    Write(LogEventLevel::ERROR, message);
}

void Logger::Fatal(std::string message) {
    Write(LogEventLevel::FATAL, message);
}

Logger* Logger::_logger = nullptr;

}

File: Simple/Logger.py
import datetime
from enum import Enum

class LogEventLevel(Enum):
    VERBO = 0
    DEBUG = 1
    INFOR = 2
    WARNN = 3
    ERROR = 4
    FATAL = 5

class LogEvent():
    def __init__(self, level, message):
        self.message = message
        self.timestamp = datetime.datetime.now()
        self.level = level

    def GetTimestamp(self):
        return self.timestamp
    
    def GetLevel(self):
        return self.level

    def GetMessage(self):
        return self.message
    

class Logger():

    def __init__(self):
        self._logPath = "Logger.log"
        self._logFile = None
        self._logFormat = "{%Y/%m/%d %H:%M:%S}: [{1}] {2}"
    
    def SetLogPath(self, log):
        self._logPath = log
             
    def Write(self, level, message):
        self.WriteImp(LogEvent(level, message))
    
    def WriteImp(self, event):
        log_string = self._logFormat.format(event.GetTimestamp(), event.GetLevel().name, event.GetMessage())
        if not self._logFile:
            self._logFile = open(self._logPath, "a+")
        self._logFile.write(log_string + "\n")
        print(log_string)

    def Verbose(self, message):
        self.Write(LogEventLevel.VERBO, message)

    def Debug(self, message):
        self.Write(LogEventLevel.DEBUG, message)

    def Information(self, message):
        self.Write(LogEventLevel.INFOR, message)

    def Warning(self, message):
        self.Write(LogEventLevel.WARNN, message)

    def Error(self, message):
        self.Write(LogEventLevel.ERROR, message)

    def Fatal(self, message):
        self.Write(LogEventLevel.FATAL, message)
